File: src/ios/VizbeeNativeManager.m
#import "VizbeeNativeManager.h"
#import "VizbeeAppAdapter.h"
#import "VizbeeVideo.h"
#import "VizbeeStyles.h"

#import <VizbeeKit/VizbeeKit.h>

@interface VizbeeNativeManager() <VZBSessionStateDelegate, VZBVideoStatusUpdateDelegate>

@property (nonatomic, assign) BOOL isSDKInitialized;

@property (nonatomic, assign) BOOL hasListeners;
@property (nonatomic, assign) VZBSessionState lastUpdatedState;

@property (nonatomic, strong) VizbeeAppAdapter* vizbeeAppAdapter;

@end

@implementation VizbeeNativeManager

//-------------
#pragma mark - Constructors & Notifications
//-------------

-(void) setChromecastAppStoreId:(CDVInvokedUrlCommand*)command {
    // Do nothing, as it is required only for Android platform to show lock/notification controls
}

-(void) initialize:(CDVInvokedUrlCommand*) command {
        
    if (!self.isSDKInitialized) {
        
        self.isSDKInitialized = YES;
        
        // get appid
        NSString* vizbeeAppId = command.arguments[0];
        NSLog(@"Initialize VizbeeSDK with AppId %@", vizbeeAppId);

        // initialize the sdk
        VZBOptions *vizbeeOptions = [VZBOptions new];
        vizbeeOptions.uiConfig = [VizbeeStyles darkTheme];
        self.vizbeeAppAdapter = [[VizbeeAppAdapter alloc] init];
        [Vizbee startWithAppID:vizbeeAppId appAdapterDelegate:self.vizbeeAppAdapter andVizbeeOptions:vizbeeOptions];

        [self initNotifications];
    } else {
        NSLog(@"Ignoring multiple calls to initialize the VizbeeSDK");
    }
}

-(void) initNotifications {

    [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(onApplicationWillResignActive:)
                                               name:UIApplicationWillResignActiveNotification
                                             object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(onApplicationDidBecomeActive:)
                                                 name:UIApplicationDidBecomeActiveNotification
                                               object:nil];
    // force first update
    [self onApplicationDidBecomeActive:nil];
}

//----------------
#pragma mark - Flow APIs
//----------------

-(void) addCastIcon:(CDVInvokedUrlCommand*) command {

    // read x and y positions
    int x = [command.arguments[0] intValue];
    int y = [command.arguments[1] intValue];
    int width = [command.arguments[2] intValue];
    int height = [command.arguments[3] intValue];
    NSLog(@"Adding CastIcon at x: %d and y: %d width: %d and height %d", x, y, width, height);
    
    [self topViewControllerThreadSafe:^(UIViewController* vc) {
        
        // create cast button
        VZBCastButton* castButton = [Vizbee createCastButton];
        castButton.castButtonType = VZBCastButtonTypeFloating;
        [vc.view addSubview:castButton];
        
        // add autolayout constraints
        castButton.translatesAutoresizingMaskIntoConstraints = false;
        NSLayoutConstraint* constraintTop = [NSLayoutConstraint constraintWithItem:castButton
                                                                         attribute:NSLayoutAttributeTop
                                                                         relatedBy:NSLayoutRelationEqual
                                                                            toItem:vc.view
                                                                         attribute:NSLayoutAttributeTop
                                                                        multiplier:1.0 constant:y];
        NSLayoutConstraint* constraintLeading = [NSLayoutConstraint constraintWithItem:castButton
                                                                             attribute:NSLayoutAttributeLeading
                                                                             relatedBy:NSLayoutRelationEqual
                                                                                toItem:vc.view
                                                                             attribute:NSLayoutAttributeLeading
                                                                            multiplier:1.0 constant:x];
        NSLayoutConstraint* constraintWidth = [NSLayoutConstraint constraintWithItem:castButton
                                                                           attribute:NSLayoutAttributeWidth
                                                                           relatedBy:NSLayoutRelationEqual
                                                                              toItem:nil
                                                                           attribute:NSLayoutAttributeNotAnAttribute
                                                                          multiplier:1.0 constant:width];
        NSLayoutConstraint* constraintHeight = [NSLayoutConstraint constraintWithItem:castButton
                                                                            attribute:NSLayoutAttributeHeight
                                                                            relatedBy:NSLayoutRelationEqual
                                                                               toItem:nil
                                                                            attribute:NSLayoutAttributeNotAnAttribute
                                                                           multiplier:1.0 constant:height];
        [NSLayoutConstraint activateConstraints:@[constraintTop, constraintLeading, constraintWidth, constraintHeight]];
    }];
}

-(void) removeCastIcon:(CDVInvokedUrlCommand *)command {

    NSLog(@"Remove CastIcon");
    [self topViewControllerThreadSafe:^(UIViewController* vc) {

        for(id view in vc.view.subviews) {

           if([view isKindOfClass:NSClassFromString(@"VZBCastButton")]) {
                NSLog(@"Found CastIcon, removing");
                [view removeFromSuperview];
           }
        }
    }];
}

-(void) smartPrompt:(CDVInvokedUrlCommand*) command {

    NSLog(@"Invoking smartPrompt");
    [self topViewControllerThreadSafe:^(UIViewController* vc) {
        
        if (nil == vc) {
            NSLog(@"SmartPrompt - nil viewcontroller");
            return;
        }
        
        // Renamed old smartHelp API to new smartPrompt
        [Vizbee smartHelp:vc];
    }];
}

-(void) smartCast:(CDVInvokedUrlCommand *)command {
    
    NSLog(@"Invoking smartCast");
    [self topViewControllerThreadSafe:^(UIViewController* vc) {
        
        if (nil == vc) {
            NSLog(@"SmartCast - nil viewcontroller");
            return;
        }
        
        VZBSessionManager* sessionManager = [Vizbee getSessionManager];
        if (nil != sessionManager) {
            [sessionManager onCastIconTapped:vc];
        }
    }];
}

-(void) smartPlay:(CDVInvokedUrlCommand *)command {
        
    self.vizbeeAppAdapter.callbackId = command.callbackId;
    self.vizbeeAppAdapter.commandDelegate = self.commandDelegate;
    
    NSLog(@"Invoking smartPlay");
    [self topViewControllerThreadSafe:^(UIViewController* vc) {
        
        if (nil == vc) {
            NSLog(@"SmartPlay - nil viewcontroller");
            
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Play on Phone"];
            [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
            return;
        }

        NSDictionary* vizbeeVideoMap = command.arguments[0];
        VizbeeVideo* vizbeeVideo = [[VizbeeVideo alloc] init:vizbeeVideoMap];
        BOOL didPlayOnTV = [Vizbee smartPlay:vizbeeVideo
                                     atPosition:vizbeeVideo.startPositionInSeconds
                       presentingViewController:vc];
                              
        if (didPlayOnTV) {

            NSLog(@"SmartPlay success in casting content");
            if (nil != command) {
                 CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Playing on TV"];
                 [pluginResult setKeepCallback:@YES];
                 [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
            }

        } else {

            NSLog(@"SmartPlay failed in casting content");
            if (nil != command){
                CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Play on Phone"];
                [pluginResult setKeepCallback:@YES];
                [self.commandDelegate sendPluginResult:pluginResult callbackId: command.callbackId];
            }
        }
    }];
}

//----------------
#pragma mark - Session APIs
//----------------
/*
RCT_REMAP_METHOD(getSessionState, getSessionStateWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    RCTLogInfo(@"Invoking getSessionState");

    VZBSessionManager* sessionManager = [Vizbee getSessionManager];
    if (nil == sessionManager) {
          reject(@"No session manager", @"Session manager is nil", nil);
          return;
    }

    resolve([self getSessionStateString:[sessionManager getSessionState]]);
}

RCT_REMAP_METHOD(getSessionConnectedDevice, getSessionConnectedDeviceWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    RCTLogInfo(@"Invoking getSessionConnectedDevice");
  
    VZBSessionManager* sessionManager = [Vizbee getSessionManager];
    if (nil == sessionManager) {
          reject(@"No session manager", @"Session manager is nil", nil);
          return;
    }
  
    NSDictionary* map = [self getSessionConnectedDeviceMap];
    RCTLogInfo(@"getSessionConnectedDevice %@", map);
    resolve(map);
}
*/

//----------------
#pragma mark - Video APIs
//----------------

-(void) play:(CDVInvokedUrlCommand *)command {

    VZBVideoClient* videoClient = [self getSessionVideoClient];
    if (nil != videoClient) {
        [videoClient play];
    } else {
        NSLog(@"Play ignored because videoClient is null");
    }
}

-(void) pause:(CDVInvokedUrlCommand *)command {

    VZBVideoClient* videoClient = [self getSessionVideoClient];
    if (nil != videoClient) {
        [videoClient pause];
    } else {
        NSLog(@"Pause ignored because videoClient is null");
    }
}

-(void) seek:(CDVInvokedUrlCommand *)command {

    VZBVideoClient* videoClient = [self getSessionVideoClient];
    NSNumber* positionNumber = command.arguments[0];
    if (nil != videoClient) {
        [videoClient seek:positionNumber.floatValue];
    } else {
        NSLog(@"Seek ignored because videoClient is null");
    }
}

-(void) stop:(CDVInvokedUrlCommand *)command {

    VZBVideoClient* videoClient = [self getSessionVideoClient];
    if (nil != videoClient) {
        [videoClient stop];
    } else {
        NSLog(@"Stop ignored because videoClient is null");
    }
}

//----------------
#pragma mark - App & session lifecycle
//----------------

-(void) onApplicationDidBecomeActive:(NSNotification*)notification {
    NSLog(@"onApplicationDidBecomeActive - adding session state listener");
    [self addSessionStateListener];
}

-(void) onApplicationWillResignActive:(NSNotification*)notification {
    NSLog(@"onApplicationWillResignActive - removing session state listener");
    [self removeSessionStateListener];
}

-(void) onSessionStateChanged:(VZBSessionState)newState {
    
    [self notifySessionStatus:newState];
    
    // handle videoClient
    if (newState == VZBSessionStateConnected) {
        [self addVideoStatusListener];
    } else {
        [self removeVideoStatusListener];
    }
}

-(void) addSessionStateListener {

    // sanity
    [self removeSessionStateListener];
    
    VZBSessionManager* sessionManager = [Vizbee getSessionManager];
    if (nil != sessionManager) {
        
        NSLog(@"CPVZBSDK: Adding session state listener");
        [sessionManager addSessionStateDelegate:self];

        // force first update
        [self notifySessionStatus:[sessionManager getSessionState]];
    }
}

-(void) removeSessionStateListener {

    VZBSessionManager* sessionManager = [Vizbee getSessionManager];
    if (nil != sessionManager) {

        NSLog(@"CPVZBSDK: Removing session state listener");
        [sessionManager removeSessionStateDelegate:self];
    }
}

-(void) notifySessionStatus:(VZBSessionState) newState {

    if (newState == self.lastUpdatedState) {
        NSLog(@"CPVZBSDK: Ignoring duplicate state update");
        return;
    }
    self.lastUpdatedState = newState;

    NSString* state = [self getSessionStateString:newState];
    NSMutableDictionary* stateMap = [NSMutableDictionary new];
    [stateMap setObject:state forKey:@"connectionState"];

    NSMutableDictionary* deviceMap = [self getSessionConnectedDeviceMap];
    if (nil != deviceMap) {
        [stateMap addEntriesFromDictionary:deviceMap];
    }

    NSLog(@"CPVZBSDK: Sending session status %@", stateMap);
//    [self sendEvent:@"VZB_SESSION_STATUS" withBody:stateMap];
}

-(NSString*) getSessionStateString:(VZBSessionState) state {

    switch (state) {
        case VZBSessionStateNoDeviceAvailable:
            return @"NO_DEVICES_AVAILABLE";
        case VZBSessionStateNotConnected:
            return @"NOT_CONNECTED";
        case VZBSessionStateConnecting:
            return @"CONNECTING";
        case VZBSessionStateConnected:
            return @"CONNECTED";
        default:
            return @"UNKNOWN";
    }
}

-(NSMutableDictionary*) getSessionConnectedDeviceMap {
    
    VZBSessionManager* sessionManager = [Vizbee getSessionManager];
    if (nil == sessionManager) {
        return nil;
    }

    VZBSession* currentSession = [sessionManager getCurrentSession];
    if (nil == currentSession) {
        return nil;
    }

    VZBScreen* screen = currentSession.vizbeeScreen;
    if (nil == screen) {
      return nil;
    }

    NSMutableDictionary* map = [NSMutableDictionary new];
    [map setObject:screen.screenType.typeName forKey:@"connectedDeviceType"];
    [map setObject:screen.screenInfo.friendlyName forKey:@"connectedDeviceFriendlyName"];
    [map setObject:screen.screenInfo.model forKey:@"connectedDeviceModel"];
    return map;
}

//----------------
#pragma mark - Video client listener
//----------------
    
- (void)onVideoStatusUpdate:(VZBVideoStatus *)videoStatus {
    [self notifyMediaStatus:videoStatus];
}

-(VZBVideoClient*) getSessionVideoClient {

    VZBSessionManager* sessionManager = [Vizbee getSessionManager];
    if (nil == sessionManager) {
        return nil;
    }

    VZBSession* currentSession = [sessionManager getCurrentSession];
    if (nil == currentSession) {
        return nil;
    }

    VZBVideoClient* videoClient = currentSession.videoClient;
    return videoClient;
}

-(void) addVideoStatusListener {

    // sanity
    [self removeVideoStatusListener];

    NSLog(@"CPVZBSDK: Trying to add video status listener");
    VZBVideoClient* videoClient = [self getSessionVideoClient];
    if (nil != videoClient) {

        [videoClient addVideoStatusDelegate:self];
        NSLog(@"CPVZBSDK: Success adding video status listener");

        // force first update
        [self notifyMediaStatus:[videoClient getVideoStatus]];
    } else {

        NSLog(@"CPVZBSDK: Failed adding video status listener");
    }
}

-(void) removeVideoStatusListener {
    
    NSLog(@"CPVZBSDK: Trying to remove video status listener");
    VZBVideoClient* videoClient = [self getSessionVideoClient];
    if (nil != videoClient) {
        
        NSLog(@"CPVZBSDK: Success removing video status listener");
        [videoClient removeVideoStatusDelegate:self];
    }
}

-(void) notifyMediaStatus:(VZBVideoStatus*) videoStatus {

    NSLog(@"CPVZBSDK: Sending media status %@", videoStatus);
//    NSDictionary* videoStatusMap = [self getVideoStatusMap:videoStatus];
//    [self sendEvent:@"VZB_MEDIA_STATUS" withBody:videoStatusMap];
}

-(NSMutableDictionary*) getVideoStatusMap:(VZBVideoStatus*) videoStatus {

    NSMutableDictionary* videoStatusMap = [NSMutableDictionary new];
    [videoStatusMap setObject:videoStatus.guid forKey:@"guid"];
    [videoStatusMap setObject:[NSNumber numberWithInt:videoStatus.streamPosition] forKey:@"position"];
    [videoStatusMap setObject:[NSNumber numberWithInt:videoStatus.streamDuration] forKey:@"duration"];
    [videoStatusMap setObject:[NSNumber numberWithBool:videoStatus.isStreamLive] forKey:@"isLive"];
    [videoStatusMap setObject:[self getPlayerStateString:videoStatus.playerState] forKey:@"playerState"];
    [videoStatusMap setObject:[NSNumber numberWithBool:videoStatus.isAdPlaying] forKey:@"isAdPlaying"];
    
    return videoStatusMap;
}

-(NSString*) getPlayerStateString:(VZBVideoPlayerState) state {

    switch (state) {
        case VZBVideoPlayerStateIdle:
            return @"IDLE";
        case VZBVideoPlayerStatePlaying:
            return @"PLAYING";
        case VZBVideoPlayerStatePaused:
            return @"PAUSED";
        case VZBVideoPlayerStateBuffering:
            return @"BUFFERING";
        default:
            return @"UNKNOWN";
    }
}

//----------------
#pragma mark - Helpers
//----------------

- (UIViewController *) topViewController {

    UIViewController *topVC = [[[UIApplication sharedApplication] keyWindow] rootViewController];
    while (topVC.presentedViewController) {
        topVC = topVC.presentedViewController;
    }
    return topVC;
}

-(void) topViewControllerThreadSafe:(void (^)(UIViewController*)) completion {
    
    dispatch_async(dispatch_get_main_queue(), ^(){
    
        UIViewController *topVC = [self topViewController];
        completion(topVC);
    });
}

@end

File: src/android/java/VizbeeNativeManager.java
package tv.vizbee.cdsender;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;

import com.google.android.gms.cast.framework.CastContext;
import com.mgm.apps.roar.R;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import tv.vizbee.api.RemoteButton;
import tv.vizbee.api.VizbeeContext;
import tv.vizbee.api.VizbeeOptions;
import tv.vizbee.api.session.SessionState;
import tv.vizbee.api.session.SessionStateListener;
import tv.vizbee.api.session.VideoClient;
import tv.vizbee.api.session.VideoStatus;
import tv.vizbee.api.session.VizbeeSession;
import tv.vizbee.api.session.VizbeeSessionManager;

public class VizbeeNativeManager extends CordovaPlugin {

    private static final String LOG_TAG = VizbeeNativeManager.class.getName();

    //----------------
    // Cordova execute
    //----------------

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext cbContext) throws JSONException {

        Log.i(LOG_TAG, "execute " + action);

        switch (action) {
            case "setChromecastAppStoreId": {
                CastOptionsProvider.receiverAppId = args.get(0).toString();
                break;
            }
            case "initialize": {
                String vizbeeAppId = args.get(0).toString();
                cordova.getActivity().runOnUiThread(() -> initialize(vizbeeAppId));
                break;
            }
            case "addCastIcon": {
                int x = (int) args.get(0);
                int y = (int) args.get(1);
                int width = (int) args.get(2);
                int heigt = (int) args.get(3);
                cordova.getActivity().runOnUiThread(() -> addCastIcon(x, y, width, heigt));
                break;
            }
            case "removeCastIcon": {
                cordova.getActivity().runOnUiThread(this::removeCastIcon);
                break;
            }
            case "smartPrompt": {
                cordova.getActivity().runOnUiThread(this::smartPrompt);
                break;
            }
            case "smartPlay": {
                JSONObject vizbeeVideoMap = (JSONObject) args.get(0);
                cordova.getActivity().runOnUiThread(() -> smartPlay(vizbeeVideoMap, cbContext));
                break;
            }
        }
        return true;
    }

    //----------------
    // Initialize
    //----------------

    public void initialize(String appId) {

        Log.v(LOG_TAG, "initialize");

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            cordova.getActivity().onTopResumedActivityChanged(true);
        }

        // init Vizbee after castContext setup for lock/notification controls
        CastContext.getSharedInstance(cordova.getActivity());

        VizbeeOptions options = new VizbeeOptions.Builder().enableProduction(true).build();
        VizbeeContext.getInstance().enableVerboseLogging();
        VizbeeContext.getInstance().init(cordova.getActivity().getApplication(), appId, new VizbeeAppAdapter(), options);
    }

    //----------------
    // UI - CastButton
    //----------------

    public void addCastIcon(int x, int y, int width, int height) {

        Log.v(LOG_TAG, "Adding CastIcon at x " + x + " y "  + y + " width " + width + " height " + height);

        LayoutInflater inflater = (LayoutInflater) cordova.getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        RemoteButton m_button = (RemoteButton) inflater.inflate(R.layout.cast_button, null);

        DisplayMetrics displayMetrics = cordova.getActivity().getResources().getDisplayMetrics();
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                (width * (int)displayMetrics.density),
                (height * (int)displayMetrics.density));
        params.leftMargin = (int) (x * displayMetrics.density);
        params.topMargin  = (int) (y * displayMetrics.density);

        FrameLayout layout = (FrameLayout) webView.getView().getParent();
        layout.addView(m_button, params);
    }

    public void removeCastIcon() {

        Log.v(LOG_TAG, "Removing CastIcon");

        FrameLayout layout = (FrameLayout) webView.getView().getParent();
        VizbeeCastButtonView vizbeeCastButtonView = null;
        for (int i = 0; i < layout.getChildCount(); i++) {
            View child = layout.getChildAt(i);
            if (child instanceof VizbeeCastButtonView) {
                vizbeeCastButtonView = (VizbeeCastButtonView) child;
                break;
            }
        }
        if (null != vizbeeCastButtonView) {
            Log.v(LOG_TAG, "Found CastIconView and removing");
            layout.removeView(vizbeeCastButtonView);
        }
    }

    //----------------
    // Flow APIs
    //----------------

    public void smartPrompt() {

        Log.v(LOG_TAG, "Invoking smartPrompt");

        Activity activity = cordova.getActivity();
        if (activity == null) {
            Log.e(LOG_TAG, "SmartPrompt - null activity");
            return;
        }

        // Renamed old smartHelp API to new smartPrompt
        VizbeeContext.getInstance().smartHelp(activity);
    }

    public void smartCast() {

        Log.v(LOG_TAG, "Invoking smartCast");

        Activity activity = cordova.getActivity();
        if (activity == null) {
            Log.e(LOG_TAG, "SmartCast - null activity");
            return;
        }

        VizbeeSessionManager sessionManager = VizbeeContext.getInstance().getSessionManager();
        if (null != sessionManager) {
            sessionManager.onCastIconClicked(activity);
        }
    }

    public void smartPlay(JSONObject vizbeeVideoMap, CallbackContext callbackContext) {

        Log.v(LOG_TAG, "Invoking smartPlay");

        Activity activity = cordova.getActivity();
        if (activity == null) {
            Log.e(LOG_TAG, "SmartPlay - null activity");
            return;
        }

        VizbeeVideo vizbeeVideo = new VizbeeVideo(vizbeeVideoMap);

        // IMPORTANT: Android expects position in milliseconds (while iOS expects in seconds!)
        boolean didPlayOnTV = VizbeeContext.getInstance().smartPlay(activity,
                vizbeeVideo, (long)(1000*vizbeeVideo.getStartPositionInSeconds()));
        if (didPlayOnTV) {

          Log.i(LOG_TAG, "SmartPlay success in casting content");
            if (callbackContext != null) {
                callbackContext.success("Playing on TV");
            }

        } else {

          Log.e(LOG_TAG, "SmartPlay failed in casting content");
            if (callbackContext != null) {
                callbackContext.success("Play on Phone");
            }
        }
    }

    //----------------
    // Session APIs
    //----------------

//    public void getSessionState(final Promise promise) {
//
//        cordova.getActivity().runOnUiQueueThread(new Runnable() {
//
//            @Override
//            public void run() {
//
//                VizbeeSessionManager sessionManager = VizbeeContext.getInstance().getSessionManager();
//                if (null == sessionManager) {
//                    promise.resolve(null);
//                    return;
//                }
//
//                promise.resolve(VizbeeNativeManager.this.getSessionStateString(sessionManager.getSessionState()));
//            }
//        });
//    }

//    public void getSessionConnectedDevice(final Promise promise) {
//
//        cordova.getActivity().runOnUiQueueThread(new Runnable() {
//
//            @Override
//            public void run() {
////                promise.resolve(VizbeeNativeManager.this.getSessionConnectedDeviceMap());
//            }
//        });
//    }

    //----------------
    // Video APIs
    //----------------

    public void play() {

        VideoClient videoClient = getSessionVideoClient();
        if (null != videoClient) {
            videoClient.play();
        } else {
            Log.w(LOG_TAG, "Play ignored because videoClient is null");
        }
    }

    public void pause() {
        
        VideoClient videoClient = getSessionVideoClient();
        if (null != videoClient) {
            videoClient.pause();
        } else {
            Log.w(LOG_TAG, "Pause ignored because videoClient is null");
        }
    }

    public void seek(long position) {
        
        VideoClient videoClient = getSessionVideoClient();
        if (null != videoClient) {
            videoClient.seek(position);
        } else {
            Log.w(LOG_TAG, "Seek ignored because videoClient is null");
        }
    }

    public void stop() {
        VideoClient videoClient = getSessionVideoClient();
        if (null != videoClient) {
            videoClient.stop();
        } else {
            Log.w(LOG_TAG, "Stop ignored because videoClient is null");
        }
    }

    //----------------
    // App & session lifecycle
    //----------------
    @Override
    public void onPause(boolean multitasking) {
        super.onPause(multitasking);

        Log.v(LOG_TAG, "onPause");
    }

    @Override
    public void onResume(boolean multitasking) {
        super.onResume(multitasking);

        Log.v(LOG_TAG, "onResume");
        this.addSessionStateListener();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        Log.v(LOG_TAG, "onDestroy");
        this.removeSessionStateListener();
    }

    private SessionStateListener sessionStateListener;
    private int lastUpdatedState = 0; // UNKNOWN

    private void addSessionStateListener() {

        // sanity
        this.removeSessionStateListener();

        VizbeeSessionManager sessionManager = VizbeeContext.getInstance().getSessionManager();
        if (null != sessionManager) {

            Log.i(LOG_TAG, "Adding session state listener");
            this.sessionStateListener = newState -> {

                // handle videoClient
                if (newState == SessionState.CONNECTED) {
                    VizbeeNativeManager.this.addVideoStatusListener();
                } else {
                    VizbeeNativeManager.this.removeVideoStatusListener();
                }

                VizbeeNativeManager.this.notifySessionStatus(newState);
            };
            sessionManager.addSessionStateListener(this.sessionStateListener);

            // force first update
            this.notifySessionStatus(sessionManager.getSessionState());
        }
    }

    private void removeSessionStateListener() {

        VizbeeSessionManager sessionManager = VizbeeContext.getInstance().getSessionManager();
        if (null != sessionManager) {
            if (null != this.sessionStateListener) {

                Log.i(LOG_TAG, "Removing session state listener");
                sessionManager.removeSessionStateListener(this.sessionStateListener);
            }
        }
        this.sessionStateListener = null;
    }

    private void notifySessionStatus(int newState) {

        if (newState == lastUpdatedState) {
            Log.w(LOG_TAG, "Ignoring duplicate state update");
            return;
        }
        lastUpdatedState = newState;

//        String state = this.getSessionStateString(newState);
//        WritableMap stateMap = Arguments.createMap();
//        stateMap.putString("connectionState", state);
//
//        WritableMap deviceMap = this.getSessionConnectedDeviceMap();
//        if (null != deviceMap) {
//            stateMap.merge(deviceMap);
//        }
//
//        this.sendEvent("VZB_SESSION_STATUS", stateMap);
    }

    private String getSessionStateString(int state) {

        switch (state) {
            case SessionState.NO_DEVICES_AVAILABLE:
                return "NO_DEVICES_AVAILABLE";
            case SessionState.NOT_CONNECTED:
                return "NOT_CONNECTED";
            case SessionState.CONNECTING:
                return "CONNECTING";
            case SessionState.CONNECTED:
                return "CONNECTED";
            default:
                return "UNKNOWN";
        }
    }

//    private WritableMap getSessionConnectedDeviceMap() {
//
//        VizbeeSessionManager sessionManager = VizbeeContext.getInstance().getSessionManager();
//        if (null == sessionManager) {
//            return null;
//        }
//
//        VizbeeSession currentSession = sessionManager.getCurrentSession();
//        if (null == currentSession) {
//            return null;
//        }
//
//        VizbeeScreen screen = currentSession.getVizbeeScreen();
//        if (null == screen) {
//          return null;
//        }
//
//        WritableMap map = Arguments.createMap();
//        map.putString("connectedDeviceType", screen.getScreenType().getTypeName());
//        map.putString("connectedDeviceFriendlyName", screen.getScreenInfo().getFriendlyName());
//        map.putString("connectedDeviceModel", screen.getScreenInfo().getModel());
//        return map;
//    }

    //----------------
    // Video client listener
    //----------------

     private VideoClient.VideoStatusListener videoStatusListener;

    private VideoClient getSessionVideoClient() {

        VizbeeSessionManager sessionManager = VizbeeContext.getInstance().getSessionManager();
        if (null == sessionManager) {
            return null;
        }

        VizbeeSession currentSession = sessionManager.getCurrentSession();
        if (null == currentSession) {
            return null;
        }

        VideoClient videoClient = currentSession.getVideoClient();
        return videoClient;
    }

    private void addVideoStatusListener() {

        // sanity
        this.removeVideoStatusListener();

        Log.v(LOG_TAG, "TRYING to add video status listener");
        VideoClient videoClient = getSessionVideoClient();
        if (null != videoClient) {

            this.videoStatusListener = new VideoClient.VideoStatusListener() {
                
                @Override
                public void onVideoStatusUpdated(VideoStatus status) {
                    VizbeeNativeManager.this.notifyMediaStatus(status);
                }
            };
            videoClient.addVideoStatusListener(this.videoStatusListener);
            Log.i(LOG_TAG, "SUCCESS adding video status listener");

            // force first update
            this.notifyMediaStatus(videoClient.getVideoStatus());
        } else {

            Log.w(LOG_TAG, "FAILED to add video status listener");
        }
    }

    private void removeVideoStatusListener() {

        Log.v(LOG_TAG, "TRYING to remove video status listener");
        VideoClient videoClient = getSessionVideoClient();
        if (null != videoClient) {
            if (null != this.videoStatusListener) {

                Log.i(LOG_TAG, "SUCCESS removing video status listener");
                videoClient.removeVideoStatusListener(this.videoStatusListener);
            }
        }
        this.videoStatusListener = null;
    }

     private void notifyMediaStatus(VideoStatus videoStatus) {

        Log.v(LOG_TAG, "Sending media status ...");
        Log.v(LOG_TAG, videoStatus.toString());
//        WritableMap videoStatusMap = this.getVideoStatusMap(videoStatus);
//        this.sendEvent("VZB_MEDIA_STATUS", videoStatusMap);
    }

//    private WritableMap getVideoStatusMap(VideoStatus videoStatus) {
//
//        WritableMap videoStatusMap = Arguments.createMap();
//        videoStatusMap.putString("guid", videoStatus.getGuid());
//        videoStatusMap.putInt("position", (int) videoStatus.getStreamPosition());
//        videoStatusMap.putInt("duration", (int) videoStatus.getStreamDuration());
//        videoStatusMap.putBoolean("isLive", videoStatus.isStreamLive());
//        videoStatusMap.putString("playerState", this.getPlayerStateString(videoStatus.getPlayerState()));
//        videoStatusMap.putBoolean("isAdPlaying", videoStatus.isAdPlaying());
//
//        return videoStatusMap;
//    }

    private String getPlayerStateString(int state) {

        switch (state) {
            case VideoStatus.PLAYER_STATE_IDLE:
                return "IDLE";
            case VideoStatus.PLAYER_STATE_PLAYING:
                return "PLAYING";
            case VideoStatus.PLAYER_STATE_PAUSED:
                return "PAUSED";
            case VideoStatus.PLAYER_STATE_BUFFERING:
                return "BUFFERING";
            default:
                return "UNKNOWN";
        }
    }

    //----------------
    // Bridge events
    //----------------

//    private void sendEvent(String eventName, @Nullable WritableMap params) {
//        getReactApplicationContext()
//                  .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
//                  .emit(eventName, params);
//    }
}